### ISSUE TITLE
Add (limited) CTFE

### ISSUE THREAD

**Issue by nrc** [rust-lang/rfcs#322]
See https://github.com/rust-lang/rfcs/pull/253 for a limited form of CTFE. We can use this issue to track a more complete form too.

**Comment by archshift** [comment]
Can we revisit this discussion now that the 1.0 launch craze has passed?

**Comment by sfackler** [comment]
This seems like it's covered by the const fn work?

**Comment by Ticki** [comment]
@sfackler Yeah, but `const fn` needs to be able to perform non-trivial functionality. For this to be done, we need `const fn` methods in traits and `impl`s of traits.

**Comment by ranma42** [comment]
Moreover the functions mentioned in #253 (`size_of`, `min_align_of`, `pref_align_of`), just like several other intrinsic-based functions, are not yet `const`.

**Comment by JinShil** [comment]
I was quite surprised that the following code does not compile:
```Rust
const CONDITION: bool = true;
const RESULT: u8 = if CONDITION { 1 } else { 0 };
```
> This seems like it's covered by the const fn work?
I thought so too, but this doesn't compile:
```Rust
#![feature(const_fn)]
pub struct BitField (u8, u8);
impl BitField
{
    const fn get_ms_bit_index(&self) -> u8 {
        if self.0 > self.1 { self.0 } else { self.1 }
    }
}
```

**Comment by nagisa** [comment]
Its in progress work. See the “postponed” label on the right hand side.

**Comment by petrochenkov** [comment]
@JinShil 
The workaround is to use an array of two elements:
```
const CONDITION: usize = 1; // true as usize
const RESULT: u8 = [0, 1][CONDITION];
```

**Comment by petrochenkov** [comment]
The lack of `if` in constant expressions is a pretty serious deficiency, its C analogue `?:` works even in preprocessor constant expressions.

**Comment by Amanieu** [comment]
If we allow `if` in constant expressions then it would make sense to allow `match` as well.

**Comment by kevincox** [comment]
Both if and match sound reasonable to me.

**Comment by frewsxcv** [comment]
Does anyone know what the status of conditionals within `const fn`s are? Is it blocked on something in particular? An RFC? Implementation?

**Comment by BenWiederhake** [comment]
Found this RFC after creating https://github.com/rust-lang/rust/issues/42906 .  What can I do to help?  What's the next step?

**Comment by eddyb** [comment]
@BenWiederhake See https://github.com/solson/miri.

**Comment by petrochenkov** [comment]
MIRI is merged into rustc and is will be used for all const evaluation soon (https://github.com/rust-lang/rust/pull/46882), and `size_of` is already constexpr, so I don't think there's value in keeping this issue open.

**Comment by jyn514** [comment]
#46882 doesn't seem to have added support for `if` in constant expressions: https://github.com/rust-lang/rust/pull/46882#issuecomment-354600638. On rust stable 1.34.0 this still gives error E019. Is this being working on?

```
error[E0019]: constant contains unimplemented expression type
 --> if.rs:3:25
  |
3 |     const number: u32 = if condition { 5 } else { 6 };
  |                         ^^^^^^^^^^
```

**Comment by Centril** [comment]
@jyn514 It's being worked on but slowly; see https://github.com/rust-lang/rust/issues/49146 and more generally https://github.com/rust-lang/rust/issues/57563 for all things `const` (`fn`).
